cmd/update.go

```diff
@@ -227,24 +232,17 @@ func (uc *updateCmd) run(cmd *cobra.Command, args []string) error {
 			return errors.Wrap(err, "failed to get VMSS list in the resource group")
 		}
 		for _, vmss := range vmssListPage.Values() {
-			vmssName := *vmss.Name
-			if sc.agentPool.OSType == api.Windows {
-				possibleIndex, nameMungingErr := strconv.Atoi(vmssName[len(vmssName)-2:])
-				if nameMungingErr != nil {
-					continue
-				}
-				if !(sc.containerService.Properties.GetAgentVMPrefix(sc.agentPool, possibleIndex) == vmssName) {
-					continue
-				}
+			vmssName := to.String(vmss.Name)
+			if sc.agentPool.VMSSName == vmssName {
+				log.Infof("found VMSS %s in resource group %s that correlates with node pool %s", vmssName, sc.resourceGroupName, sc.agentPoolToScale)
 			} else {
-				if !sc.vmInAgentPool(vmssName, vmss.Tags) {
-					continue
-				}
+				continue
 			}
 
 			if vmss.Sku != nil {
 				sc.newDesiredAgentCount = int(*vmss.Sku.Capacity)
 				uc.agentPool.Count = sc.newDesiredAgentCount
+				break
```

Comment:
Why is the additional break needed here and wasn't before?  Looks like it will stop processing and kick you out of the for loop.

Reply:
Right, so basically the current implementation enumerates through every VMSS *no matter what*. Even after it finds the VMSS that matches the pool in the API model and updates the scale command instance to reflect the current VMSS capacity (because `aks-engine update` is a `aks-engine scale` under the hood that does *not* change the instance count) and updates the API model (a convenience to reflect in the apimodel.JSON what the new count is at the time we run this operation), we keep enumerating through the VMSS array if there are more.

There is always only going to be one matching VMSS, so once we find it, we *should* break out of the enumeration and not waste any more cycles.

So yeah, this is just us improving things since we're mucking around in this area anyways.

Hope that makes sense. (Also definitely check that my reasoning is sound!)

Reply:
sounds about right just wanted to make sure it was intentional :-)

Are there other checks in place that affirm `always only going to be one matching VMSS`?  Is there any chance that something could go wrong and there are two which would lead to strange behavior?

Reply:
It does look like scale is equivalently inefficient:

https://github.com/Azure/aks-engine/blob/release-v0.58.0/cmd/scale.go#L410

Reply:
Azure VMSS API will guarantee uniquess of VMSS names in a resource group; here's how we enforce that in aks-engine:

https://github.com/Azure/aks-engine/blob/release-v0.58.0/pkg/api/vlabs/validate.go#L516